workbench/network/common/dhcp/relay/dhcrelay.8
.\"	dhcrelay.8
.\"
.\" Copyright (c) 2004 by Internet Systems Consortium, Inc. ("ISC")
.\" Copyright (c) 1997-2003 by Internet Software Consortium
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS.  IN NO EVENT SHALL ISC BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
.\" OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\"   Internet Systems Consortium, Inc.
.\"   950 Charter Street
.\"   Redwood City, CA 94063
.\"   <info@isc.org>
.\"   http://www.isc.org/
.\"
.\" This software has been written for Internet Systems Consortium
.\" by Ted Lemon in cooperation with Vixie
.\" Enterprises.  To learn more about Internet Systems Consortium,
.\" see ``http://www.isc.org/isc''.  To learn more about Vixie
.\" Enterprises, see ``http://www.vix.com''.
.\"
.\" $Id: dhcrelay.8,v 1.1.1.1 2005/12/07 10:50:33 sonic_amiga Exp $
.\"
.TH dhcrelay 8
.SH NAME
dhcrelay - Dynamic Host Configuration Protocol Relay Agent
.SH SYNOPSIS
.B dhcrelay
[
.B -p
.I port
]
[
.B -d
]
[
.B -q
]
[
.B -i
.I if0
[
.B ...
.B -i
.I ifN
]
]
[
.B -a
]
[
.B -c
.I count
]
[
.B -A
.I length
]
[
.B -D
]
[
.B -m
.I append
|
.I replace
|
.I forward
|
.I discard
]
.I server0
[
.I ...serverN
]
.SH DESCRIPTION
The Internet Systems Consortium DHCP Relay Agent, dhcrelay, provides a
means for relaying DHCP and BOOTP requests from a subnet to which
no DHCP server is directly connected to one or more DHCP servers on other
subnets.
.SH OPERATION
.PP
The DHCP Relay Agent listens for DHCP and BOOTP queries and responses.
When a query is received from a client, dhcrelay forwards it to the
list of DHCP servers specified on the command line.  When a reply is
received from a server, it is broadcast or unicast (according to the
relay agent's ability or the client's request) on the network from
which the original request came.
.SH COMMAND LINE
.PP
The names of the network interfaces that dhcrelay should attempt to
configure may be specified on the command line using the
.B -i
option.  If no interface names
are specified on the command line dhcrelay will identify all network
interfaces, elimininating non-broadcast interfaces if possible, and
attempt to configure each interface.
.PP
The
.B -i
flag can be used to specify the network interfaces on which the relay
agent should listen.   In general, it must listen not only on those
network interfaces to which clients are attached, but also on those
network interfaces to which the server (or the router that reaches the
server) is attached.   However, in some cases it may be necessary to
exclude some networks; in this case, you must list all those network
interfaces that should \fInot\fR be excluded using the \fB-i\fR flag.
.PP
In some cases it
.I is
helpful for the relay agent to forward requests from networks on which
a DHCP server is running to other DHCP servers.   This would be the
case if two DHCP servers on different networks were being used to
provide backup service for each other's networks.
.PP
If dhcrelay should listen and transmit on a port other than the
standard (port 67), the
.B -p
flag may used.  It should be followed by the udp port number that
dhcrelay should use.  This is mostly useful for debugging purposes.
.PP
Dhcrelay will normally run in the foreground until it has configured
an interface, and then will revert to running in the background.
To force dhcrelay to always run as a foreground process, the
.B -d
flag should be specified.  This is useful when running dhcrelay under
a debugger, or when running it out of inittab on System V systems.
.PP
Dhcrelay will normally print its network configuration on startup.
This can be unhelpful in a system startup script - to disable this
behaviour, specify the
.B -q
flag.
.SH RELAY AGENT INFORMATION OPTIONS
If the
.B -a
flag is set the relay agent will append an agent option field to each
request before forwarding it to the server.   Agent option fields in
responses sent from servers to clients will be stripped before
forwarding such responses back to the client.
.PP
The agent option field will contain two agent options: the Circuit ID
suboption and the Remote ID suboption.  Currently, the Circuit ID will
be the printable name of the interface on which the client request was
received.  The client supports inclusion of a Remote ID suboption as
well, but this is not used by default.
.PP
When forwarding packets, dhcrelay discards packets which have reached a hop
count of 10.  If a lower or higher threshold (up to 255) is desired, depending
on your environment, you can specify the max hop count threshold as a number
following the
.B -c
option.
.PP
Relay Agent options are added to a DHCP packet without the knowledge
of the DHCP client.   The client may have filled the DHCP packet
option buffer completely, in which case there theoretically isn't any
space to add Agent options.   However, the DHCP server may be able to
handle a much larger packet than most DHCP clients would send.   The
current Agent Options draft requires that the relay agent use a
maximum packet size of 576 bytes.   
.PP
It is recommended that with the Internet Systems Consortium DHCP
server, the maximum packet size be set to about 1400, allowing plenty
of extra space in which the relay agent can put the agent option
field, while still fitting into the Ethernet MTU size.  This can be
done by specifying the
.B -A
flag, followed by the desired maximum packet size (e.g., 1400).
.PP
Note that this is reasonably safe to do even if the MTU between the
server and the client is less than 1500, as long as the hosts on which
the server and client are running support IP fragmentation (and they
should).  With some knowledge as to how large the agent options might
get in a particular configuration, this parameter can be tuned as
finely as necessary.
.PP
It is possible for a relay agent to receive a packet which already
contains an agent option field.  If this packet does not have a giaddr
set, the standard requires that the packet be discarded.
.PP
If giaddr is set, the server may handle the situation in one of four
ways: it may
.I append
its own set of relay options to the packet, leaving the
supplied option field intact.   It may
.I replace
the existing agent option field.
It may
.I forward
the packet unchanged.   Or, it may
.I discard
it.
.PP
Which of these behaviours is followed by the Internet Systems
Consortium DHCP Relay Agent may be configured with the
.B -m
flag, followed by one of the four keywords specified in
.I italics
above.
.PP
When the relay agent receives a reply from a server that it's supposed
to forward to a client, and Relay Agent Information option processing
is enabled, the relay agent scans the packet for Relay Agent
Information options and removes them.   As it's scanning, if it finds
a Relay Agent Information option field containing an Agent ID
suboption that matches one of its IP addresses, that option is
recognized as its own.   If no such option is found, the relay agent
can either drop the packet, or relay it anyway.   If the
.B -D
option is specified, all packets that don't contain a match will be
dropped.
.SH SPECIFYING DHCP SERVERS
The name or IP address of at least one DHCP server to which DHCP and
BOOTP requests should be relayed must be specified on the command
line.
.SH SEE ALSO
dhclient(8), dhcpd(8), RFC2132, RFC2131, draft-ietf-dhc-agent-options-03.txt.
.SH BUGS
It should be possible for the user to define the Circuit ID and Remote
ID values on a per-interface basis.
.PP
The relay agent should not relay packets received on a physical
network to DHCP servers on the same physical network - if they do, the
server will receive duplicate packets.   In order to fix this,
however, the relay agent needs to be able to learn about the network
topology, which requires that it have a configuration file.
.SH AUTHOR
.B dhcrelay(8)
has been written for Internet Systems Consortium
by Ted Lemon in cooperation with Vixie
Enterprises.  To learn more about Internet Systems Consortium,
see
.B http://www.isc.org/isc.
To learn more about Vixie
Enterprises, see
.B http://www.vix.com.


workbench/network/common/dhcp/relay/dhcrelay.c
/* dhcrelay.c

   DHCP/BOOTP Relay Agent. */

/*
 * Copyright (c) 2004-2005 by Internet Systems Consortium, Inc. ("ISC")
 * Copyright (c) 1997-2003 by Internet Software Consortium
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS.  IN NO EVENT SHALL ISC BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
 * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *   Internet Systems Consortium, Inc.
 *   950 Charter Street
 *   Redwood City, CA 94063
 *   <info@isc.org>
 *   http://www.isc.org/
 *
 * This software has been written for Internet Systems Consortium
 * by Ted Lemon in cooperation with Vixie Enterprises and Nominum, Inc.
 * To learn more about Internet Systems Consortium, see
 * ``http://www.isc.org/''.  To learn more about Vixie Enterprises,
 * see ``http://www.vix.com''.   To learn more about Nominum, Inc., see
 * ``http://www.nominum.com''.
 */

#ifndef lint
static char ocopyright[] =
"$Id$ Copyright (c) 2004-2005 Internet Systems Consortium.  All rights reserved.\n";
#endif /* not lint */

#include "dhcpd.h"
#include "version.h"

static void usage PROTO ((void));

TIME default_lease_time = 43200; /* 12 hours... */
TIME max_lease_time = 86400; /* 24 hours... */
struct tree_cache *global_options [256];

/* Needed to prevent linking against conflex.c. */
int lexline;
int lexchar;
char *token_line;
char *tlname;

const char *path_dhcrelay_pid = _PATH_DHCRELAY_PID;

int bogus_agent_drops = 0;	/* Packets dropped because agent option
				   field was specified and we're not relaying
				   packets that already have an agent option
				   specified. */
int bogus_giaddr_drops = 0;	/* Packets sent to us to relay back to a
				   client, but with a bogus giaddr. */
int client_packets_relayed = 0;	/* Packets relayed from client to server. */
int server_packet_errors = 0;	/* Errors sending packets to servers. */
int server_packets_relayed = 0;	/* Packets relayed from server to client. */
int client_packet_errors = 0;	/* Errors sending packets to clients. */

int add_agent_options = 0;	/* If nonzero, add relay agent options. */
int drop_agent_mismatches = 0;	/* If nonzero, drop server replies that
				   don't have matching circuit-id's. */
int corrupt_agent_options = 0;	/* Number of packets dropped because
				   relay agent information option was bad. */
int missing_agent_option = 0;	/* Number of packets dropped because no
				   RAI option matching our ID was found. */
int bad_circuit_id = 0;		/* Circuit ID option in matching RAI option
				   did not match any known circuit ID. */
int missing_circuit_id = 0;	/* Circuit ID option in matching RAI option
				   was missing. */
int max_hop_count = 10;		/* Maximum hop count */


	/* Maximum size of a packet with agent options added. */
int dhcp_max_agent_option_packet_length = 576;

	/* What to do about packets we're asked to relay that
	   already have a relay option: */
enum { forward_and_append,	/* Forward and append our own relay option. */
       forward_and_replace,	/* Forward, but replace theirs with ours. */
       forward_untouched,	/* Forward without changes. */
       discard } agent_relay_mode = forward_and_replace;

u_int16_t local_port;
u_int16_t remote_port;

struct server_list {
	struct server_list *next;
	struct sockaddr_in to;
} *servers;

static char copyright [] = "Copyright 2004-2005 Internet Systems Consortium.";
static char arr [] = "All rights reserved.";
static char message [] = "Internet Systems Consortium DHCP Relay Agent";
static char url [] = "For info, please visit http://www.isc.org/sw/dhcp/";

int main (argc, argv, envp)
	int argc;
	char **argv, **envp;
{
	int i;
	struct servent *ent;
	struct server_list *sp = (struct server_list *)0;
	int no_daemon = 0;
	int quiet = 0;
	isc_result_t status;
	char *s;

	/* Make sure we have stdin, stdout and stderr. */
#ifdef AMIGA
	struct Process *me;
	
	me = (struct Process *)FindTask(NULL);
	if (!me->pr_COS)
		log_perror = 0;
#else
	i = open ("/dev/null", O_RDWR);
	if (i == 0)
		i = open ("/dev/null", O_RDWR);
	if (i == 1) {
		i = open ("/dev/null", O_RDWR);
		log_perror = 0; /* No sense logging to /dev/null. */
	} else if (i != -1)
		close (i);
#endif
#ifdef SYSLOG_4_2
	openlog ("dhcrelay", LOG_NDELAY);
	log_priority = LOG_DAEMON;
#else
	openlog ("dhcrelay", LOG_NDELAY, LOG_DAEMON);
#endif

#if !(defined (DEBUG) || defined (SYSLOG_4_2))
	setlogmask (LOG_UPTO (LOG_INFO));
#endif	

	/* Set up the OMAPI. */
	status = omapi_init ();
	if (status != ISC_R_SUCCESS)
		log_fatal ("Can't initialize OMAPI: %s",
			   isc_result_totext (status));

	/* Set up the OMAPI wrappers for the interface object. */
	interface_setup ();

	for (i = 1; i < argc; i++) {
		if (!strcmp (argv [i], "-p")) {
			if (++i == argc)
				usage ();
			local_port = htons (atoi (argv [i]));
			log_debug ("binding to user-specified port %d",
			       ntohs (local_port));
		} else if (!strcmp (argv [i], "-d")) {
			no_daemon = 1;
 		} else if (!strcmp (argv [i], "-i")) {
			struct interface_info *tmp =
				(struct interface_info *)0;
			status = interface_allocate (&tmp, MDL);
			if (status != ISC_R_SUCCESS)
				log_fatal ("%s: interface_allocate: %s",
					   argv [i],
					   isc_result_totext (status));
			if (++i == argc) {
				usage ();
			}
			strcpy (tmp -> name, argv [i]);
			interface_snorf (tmp, INTERFACE_REQUESTED);
			interface_dereference (&tmp, MDL);
		} else if (!strcmp (argv [i], "-q")) {
			quiet = 1;
			quiet_interface_discovery = 1;
		} else if (!strcmp (argv [i], "-a")) {
			add_agent_options = 1;
		} else if (!strcmp (argv [i], "-c")) {
			int hcount;
			if (++i == argc)
				usage ();
			hcount = atoi(argv[i]);
			if (hcount <= 255)
				max_hop_count= hcount;
			else
				usage ();
		} else if (!strcmp (argv [i], "-A")) {
			if (++i == argc)
				usage ();
			dhcp_max_agent_option_packet_length = atoi (argv [i]);
		} else if (!strcmp (argv [i], "-m")) {
			if (++i == argc)
				usage ();
			if (!strcasecmp (argv [i], "append")) {
				agent_relay_mode = forward_and_append;
			} else if (!strcasecmp (argv [i], "replace")) {
				agent_relay_mode = forward_and_replace;
			} else if (!strcasecmp (argv [i], "forward")) {
				agent_relay_mode = forward_untouched;
			} else if (!strcasecmp (argv [i], "discard")) {
				agent_relay_mode = discard;
			} else
				usage ();
		} else if (!strcmp (argv [i], "-D")) {
			drop_agent_mismatches = 1;
 		} else if (argv [i][0] == '-') {
 		    usage ();
		} else if (!strcmp (argv [i], "--version")) {
			log_info ("isc-dhcrelay-%s", DHCP_VERSION);
			exit (0);
 		} else {
			struct hostent *he;
			struct in_addr ia, *iap = (struct in_addr *)0;
			if (inet_aton (argv [i], &ia)) {
				iap = &ia;
			} else {
				he = gethostbyname (argv [i]);
				if (!he) {
					log_error ("%s: host unknown",
						   argv [i]);
				} else {
					iap = ((struct in_addr *)
					       he -> h_addr_list [0]);
				}
			}
			if (iap) {
				sp = ((struct server_list *)
				      dmalloc (sizeof *sp, MDL));
				if (!sp)
					log_fatal ("no memory for server.\n");
				sp -> next = servers;
				servers = sp;
				memcpy (&sp -> to.sin_addr,
					iap, sizeof *iap);
			}
 		}
	}

	if ((s = getenv ("PATH_DHCRELAY_PID"))) {
		path_dhcrelay_pid = s;
	}

	if (!quiet) {
		log_info ("%s %s", message, DHCP_VERSION);
		log_info (copyright);
		log_info (arr);
		log_info (url);
	} else {
		quiet = 0;
		log_perror = 0;
	}

	/* Default to the DHCP/BOOTP port. */
	if (!local_port) {
		ent = getservbyname ("dhcps", "udp");
		if (!ent)
			local_port = htons (67);
		else
			local_port = ent -> s_port;
		endservent ();
	}
	remote_port = htons (ntohs (local_port) + 1);
  
	/* We need at least one server. */
	if (!sp) {
		usage ();
	}

	/* Set up the server sockaddrs. */
	for (sp = servers; sp; sp = sp -> next) {
		sp -> to.sin_port = local_port;
		sp -> to.sin_family = AF_INET;
#ifdef HAVE_SA_LEN
		sp -> to.sin_len = sizeof sp -> to;
#endif
	}

	/* Get the current time... */
	GET_TIME (&cur_time);

	/* Discover all the network interfaces. */
	discover_interfaces (DISCOVER_RELAY);

	/* Set up the bootp packet handler... */
	bootp_packet_handler = relay;

	/* Become a daemon... */
#ifndef FORK_MISSING
	if (!no_daemon) {
		int pid;
		FILE *pf;
		int pfdesc;

		log_perror = 0;

		if ((pid = fork()) < 0)
			log_fatal ("can't fork daemon: %m");
		else if (pid)
			exit (0);

		pfdesc = open (path_dhcrelay_pid,
			       O_CREAT | O_TRUNC | O_WRONLY, 0644);

		if (pfdesc < 0) {
			log_error ("Can't create %s: %m", path_dhcrelay_pid);
		} else {
			pf = fdopen (pfdesc, "w");
			if (!pf)
				log_error ("Can't fdopen %s: %m",
				      path_dhcrelay_pid);
			else {
				fprintf (pf, "%ld\n", (long)getpid ());
				fclose (pf);
			}	
		}

		close (0);
		close (1);
		close (2);
		pid = setsid ();
	}
#endif
	/* Start dispatching packets and timeouts... */
	dispatch ();

	/*NOTREACHED*/
	return 0;
}

void relay (ip, packet, length, from_port, from, hfrom)
	struct interface_info *ip;
	struct dhcp_packet *packet;
	unsigned length;
	unsigned int from_port;
	struct iaddr from;
	struct hardware *hfrom;
{
	struct server_list *sp;
	struct sockaddr_in to;
	struct interface_info *out;
	struct hardware hto, *htop;

	if (packet -> hlen > sizeof packet -> chaddr) {
		log_info ("Discarding packet with invalid hlen.");
		return;
	}

	/* Find the interface that corresponds to the giaddr
	   in the packet. */
	if (packet -> giaddr.s_addr) {
		for (out = interfaces; out; out = out -> next) {
			if (!memcmp (&out -> primary_address,
				     &packet -> giaddr,
				     sizeof packet -> giaddr))
				break;
		}
	} else {
		out = (struct interface_info *)0;
	}

	/* If it's a bootreply, forward it to the client. */
	if (packet -> op == BOOTREPLY) {
		if (!(packet -> flags & htons (BOOTP_BROADCAST)) &&
			can_unicast_without_arp (out)) {
			to.sin_addr = packet -> yiaddr;
			to.sin_port = remote_port;

			/* and hardware address is not broadcast */
			htop = &hto;
		} else {
			to.sin_addr.s_addr = htonl (INADDR_BROADCAST);
			to.sin_port = remote_port;

			/* hardware address is broadcast */
			htop = NULL;
		}
		to.sin_family = AF_INET;
#ifdef HAVE_SA_LEN
		to.sin_len = sizeof to;
#endif

		memcpy (&hto.hbuf [1], packet -> chaddr, packet -> hlen);
		hto.hbuf [0] = packet -> htype;
		hto.hlen = packet -> hlen + 1;

		/* Wipe out the agent relay options and, if possible, figure
		   out which interface to use based on the contents of the
		   option that we put on the request to which the server is
		   replying. */
		if (!(length =
		      strip_relay_agent_options (ip, &out, packet, length)))
			return;

		if (!out) {
			log_error ("packet to bogus giaddr %s.\n",
			      inet_ntoa (packet -> giaddr));
			++bogus_giaddr_drops;
			return;
		}

		if (send_packet (out,
				 (struct packet *)0,
				 packet, length, out -> primary_address,
				 &to, htop) < 0) {
			++server_packet_errors;
		} else {
			log_debug ("forwarded BOOTREPLY for %s to %s",
			       print_hw_addr (packet -> htype, packet -> hlen,
					      packet -> chaddr),
			       inet_ntoa (to.sin_addr));

			++server_packets_relayed;
		}
		return;
	}

	/* If giaddr matches one of our addresses, ignore the packet -
	   we just sent it. */
	if (out)
		return;

	/* Add relay agent options if indicated.   If something goes wrong,
	   drop the packet. */
	if (!(length = add_relay_agent_options (ip, packet, length,
						ip -> primary_address)))
		return;

	/* If giaddr is not already set, Set it so the server can
	   figure out what net it's from and so that we can later
	   forward the response to the correct net.    If it's already
	   set, the response will be sent directly to the relay agent
	   that set giaddr, so we won't see it. */
	if (!packet -> giaddr.s_addr)
		packet -> giaddr = ip -> primary_address;
	if (packet -> hops < max_hop_count)
		packet -> hops = packet -> hops + 1;
	else
		return;

	/* Otherwise, it's a BOOTREQUEST, so forward it to all the
	   servers. */
	for (sp = servers; sp; sp = sp -> next) {
		if (send_packet ((fallback_interface
				  ? fallback_interface : interfaces),
				 (struct packet *)0,
				 packet, length, ip -> primary_address,
				 &sp -> to, (struct hardware *)0) < 0) {
			++client_packet_errors;
		} else {
			log_debug ("forwarded BOOTREQUEST for %s to %s",
			       print_hw_addr (packet -> htype, packet -> hlen,
					      packet -> chaddr),
			       inet_ntoa (sp -> to.sin_addr));
			++client_packets_relayed;
		}
	}
				 
}

static void usage ()
{
	log_fatal ("Usage: dhcrelay [-p <port>] [-d] [-D] [-i %s%s%s%s",
		"interface] [-q] [-a]\n                ",
		"[-c count] [-A length] ",
		"[-m append|replace|forward|discard]\n",
		"                [server1 [... serverN]]");
}

int write_lease (lease)
	struct lease *lease;
{
	return 1;
}

int write_host (host)
	struct host_decl *host;
{
	return 1;
}

int commit_leases ()
{
	return 1;
}

void bootp (packet)
	struct packet *packet;
{
}

void dhcp (packet)
	struct packet *packet;
{
}

int find_subnet (struct subnet **sp,
		 struct iaddr addr, const char *file, int line)
{
	return 0;
}

#if defined (DEBUG)
int check_collection (struct packet *p, struct lease *l,
		      struct collection *c)
{
	return 0;
}

void classify (struct packet *p, struct class *c)
{
}

isc_result_t find_class (struct class **class, const char *c1,
			 const char *c2, int i)
{
	return ISC_R_NOTFOUND;
}

int parse_allow_deny (struct option_cache **oc, struct parse *p, int i)
{
	return 0;
}

/* As a wise man once said in dhcpctl/omshell.c: */
/* Sigh */
isc_result_t dhcp_set_control_state (control_object_state_t oldstate,
				     control_object_state_t newstate)
{
	return ISC_R_SUCCESS;
}

#endif

/* Strip any Relay Agent Information options from the DHCP packet
   option buffer.   If there is a circuit ID suboption, look up the
   outgoing interface based upon it. */

int strip_relay_agent_options (in, out, packet, length)
	struct interface_info *in, **out;
	struct dhcp_packet *packet;
	unsigned length;
{
	int is_dhcp = 0;
	u_int8_t *op, *sp, *max;
	int good_agent_option = 0;
	int status;

	/* If we're not adding agent options to packets, we're not taking
	   them out either. */
	if (!add_agent_options)
		return length;

	/* If there's no cookie, it's a bootp packet, so we should just
	   forward it unchanged. */
	if (memcmp (packet -> options, DHCP_OPTIONS_COOKIE, 4))
		return length;

	max = ((u_int8_t *)packet) + length;
	sp = op = &packet -> options [4];

	while (op < max) {
		switch (*op) {
			/* Skip padding... */
		      case DHO_PAD:
			if (sp != op)
				*sp = *op;
			++op;
			++sp;
			continue;

			/* If we see a message type, it's a DHCP packet. */
		      case DHO_DHCP_MESSAGE_TYPE:
			is_dhcp = 1;
			goto skip;
			break;

			/* Quit immediately if we hit an End option. */
		      case DHO_END:
			if (sp != op)
				*sp++ = *op++;
			goto out;

		      case DHO_DHCP_AGENT_OPTIONS:
			/* We shouldn't see a relay agent option in a
			   packet before we've seen the DHCP packet type,
			   but if we do, we have to leave it alone. */
			if (!is_dhcp)
				goto skip;

			status = find_interface_by_agent_option (packet,
								 out, op + 2,
								 op [1]);
			if (status == -1 && drop_agent_mismatches)
				return 0;
			if (status)
				good_agent_option = 1;
			op += op [1] + 2;
			break;

		      skip:
			/* Skip over other options. */
		      default:
			if (sp != op)
				memcpy (sp, op, (unsigned)(op [1] + 2));
			sp += op [1] + 2;
			op += op [1] + 2;
			break;
		}
	}
      out:

	/* If it's not a DHCP packet, we're not supposed to touch it. */
	if (!is_dhcp)
		return length;

	/* If none of the agent options we found matched, or if we didn't
	   find any agent options, count this packet as not having any
	   matching agent options, and if we're relying on agent options
	   to determine the outgoing interface, drop the packet. */

	if (!good_agent_option) {
		++missing_agent_option;
		if (drop_agent_mismatches)
			return 0;
	}

	/* Adjust the length... */
	if (sp != op) {
		length = sp - ((u_int8_t *)packet);

		/* Make sure the packet isn't short (this is unlikely,
                   but WTH) */
		if (length < BOOTP_MIN_LEN) {
			memset (sp, 0, BOOTP_MIN_LEN - length);
			length = BOOTP_MIN_LEN;
		}
	}
	return length;
}


/* Find an interface that matches the circuit ID specified in the
   Relay Agent Information option.   If one is found, store it through
   the pointer given; otherwise, leave the existing pointer alone.

   We actually deviate somewhat from the current specification here:
   if the option buffer is corrupt, we suggest that the caller not
   respond to this packet.  If the circuit ID doesn't match any known
   interface, we suggest that the caller to drop the packet.  Only if
   we find a circuit ID that matches an existing interface do we tell
   the caller to go ahead and process the packet. */

int find_interface_by_agent_option (packet, out, buf, len)
	struct dhcp_packet *packet;
	struct interface_info **out;
	u_int8_t *buf;
	int len;
{
	int i = 0;
	u_int8_t *circuit_id = 0;
	unsigned circuit_id_len = 0;
	struct interface_info *ip;

	while (i < len) {
		/* If the next agent option overflows the end of the
		   packet, the agent option buffer is corrupt. */
		if (i + 1 == len ||
		    i + buf [i + 1] + 2 > len) {
			++corrupt_agent_options;
			return -1;
		}
		switch (buf [i]) {
			/* Remember where the circuit ID is... */
		      case RAI_CIRCUIT_ID:
			circuit_id = &buf [i + 2];
			circuit_id_len = buf [i + 1];
			i += circuit_id_len + 2;
			continue;

		      default:
			i += buf [i + 1] + 2;
			break;
		}
	}

	/* If there's no circuit ID, it's not really ours, tell the caller
	   it's no good. */
	if (!circuit_id) {
		++missing_circuit_id;
		return -1;
	}

	/* Scan the interface list looking for an interface whose
	   name matches the one specified in circuit_id. */

	for (ip = interfaces; ip; ip = ip -> next) {
		if (ip -> circuit_id &&
		    ip -> circuit_id_len == circuit_id_len &&
		    !memcmp (ip -> circuit_id, circuit_id, circuit_id_len))
			break;
	}

	/* If we got a match, use it. */
	if (ip) {
		*out = ip;
		return 1;
	}

	/* If we didn't get a match, the circuit ID was bogus. */
	++bad_circuit_id;
	return -1;
}

/* Examine a packet to see if it's a candidate to have a Relay
   Agent Information option tacked onto its tail.   If it is, tack
   the option on.  */

int add_relay_agent_options (ip, packet, length, giaddr)
	struct interface_info *ip;
	struct dhcp_packet *packet;
	unsigned length;
	struct in_addr giaddr;
{
	int is_dhcp = 0, agent_options_present = 0;
	u_int8_t *op, *sp, *max, *end_pad = 0;

	/* If we're not adding agent options to packets, we can skip
	   this. */
	if (!add_agent_options)
		return length;

	/* If there's no cookie, it's a bootp packet, so we should just
	   forward it unchanged. */
	if (memcmp (packet -> options, DHCP_OPTIONS_COOKIE, 4))
		return length;

	max = ((u_int8_t *)packet) + length;
	sp = op = &packet -> options [4];

	while (op < max) {
		switch (*op) {
			/* Skip padding... */
		      case DHO_PAD:
			end_pad = sp;
			if (sp != op)
				*sp = *op;
			++op;
			++sp;
			continue;

			/* If we see a message type, it's a DHCP packet. */
		      case DHO_DHCP_MESSAGE_TYPE:
			is_dhcp = 1;
			goto skip;
			break;

			/* Quit immediately if we hit an End option. */
		      case DHO_END:
			goto out;

		      case DHO_DHCP_AGENT_OPTIONS:
			/* We shouldn't see a relay agent option in a
			   packet before we've seen the DHCP packet type,
			   but if we do, we have to leave it alone. */
			if (!is_dhcp)
				goto skip;
			end_pad = 0;

			/* There's already a Relay Agent Information option
			   in this packet.   How embarrassing.   Decide what
			   to do based on the mode the user specified. */

			switch (agent_relay_mode) {
			      case forward_and_append:
				goto skip;
			      case forward_untouched:
				return length;
			      case discard:
				return 0;
			      case forward_and_replace:
			      default:
				break;
			}

			/* Skip over the agent option and start copying
			   if we aren't copying already. */
			op += op [1] + 2;
			break;

		      skip:
			/* Skip over other options. */
		      default:
			end_pad = 0;
			if (sp != op)
				memcpy (sp, op, (unsigned)(op [1] + 2));
			sp += op [1] + 2;
			op += op [1] + 2;
			break;
		}
	}
      out:

	/* If it's not a DHCP packet, we're not supposed to touch it. */
	if (!is_dhcp)
		return length;

	/* If the packet was padded out, we can store the agent option
	   at the beginning of the padding. */

	if (end_pad)
		sp = end_pad;

	/* Remember where the end of the packet was after parsing
	   it. */
	op = sp;

	/* XXX Is there room? */

	/* Okay, cons up *our* Relay Agent Information option. */
	*sp++ = DHO_DHCP_AGENT_OPTIONS;
	*sp++ = 0;	/* Dunno... */

	/* Copy in the circuit id... */
	*sp++ = RAI_CIRCUIT_ID;
	/* Sanity check.   Had better not every happen. */
	if (ip -> circuit_id_len > 255 || ip -> circuit_id_len < 1)
		log_fatal ("completely bogus circuit id length %d on %s\n",
		       ip -> circuit_id_len, ip -> name);
	*sp++ = ip -> circuit_id_len;
	memcpy (sp, ip -> circuit_id, ip -> circuit_id_len);
	sp += ip -> circuit_id_len;

	/* Copy in remote ID... */
	if (ip -> remote_id) {
		*sp++ = RAI_REMOTE_ID;
		if (ip -> remote_id_len > 255 || ip -> remote_id_len < 1)
			log_fatal ("bogus remote id length %d on %s\n",
			       ip -> circuit_id_len, ip -> name);
		*sp++ = ip -> remote_id_len;
		memcpy (sp, ip -> remote_id, ip -> remote_id_len);
		sp += ip -> remote_id_len;
	}

	/* Relay option's total length shouldn't ever get to be more than
	   257 bytes. */
	if (sp - op > 257)
	    log_fatal ("total agent option length exceeds 257 (%ld) on %s\n",
		       (long)(sp - op), ip -> name);

	/* Calculate length of RAI option. */
	op [1] = sp - op - 2;

	/* Deposit an END token. */
	*sp++ = DHO_END;

	/* Recalculate total packet length. */
	length = sp - ((u_int8_t *)packet);

	/* Make sure the packet isn't short (this is unlikely, but WTH) */
	if (length < BOOTP_MIN_LEN) {
		memset (sp, 0, BOOTP_MIN_LEN - length);
		length = BOOTP_MIN_LEN;
	}

	return length;
}
